# Exploring Hacker News Posts

## What is Hacker News? 
Hacker News is a social news site founded by the startup incubator Y Combinator. Users submit posts which are voted and commented on. Topics of posts are usually focused on computer science and entreprenuership and the site is popular in technology and startup circles. 

Sources: DataQuest and Wikipedia

## Data Source
The data is available from [Kaggle](https://www.kaggle.com/hacker-news/hacker-news-posts)
However, the dataset provided by DataQuest has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining sumbissions. 

**Data Dictionary**
- id: The unique identifier from Hacker News for the post
- title: The title of the post
- url: The URL that the posts links to, if it the post has a URL
- num_points: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
- num_comments: The number of comments that were made on the post
- author: The username of the person who submitted the post
- created_at: The date and time at which the post was submitted

Source: Dataquest

## How Hacker News Works
Posts can be up voted, down voted, and/or commented on. Posts show the Title, number of points, author's user name, how long ago it was posted, and the number of comments. Hacker News has several tabs including:
 - new
 - past
 - comments
 - ask
 - show
These allow the reader to click and explore Hacker News based on the user's preference. Ask and Show are features of Hacker News where the poster includes Ask HN or Show HN in the title. Ask HN is where a user can ask Hacker News users a question. Show HN is where a user can show Hacker News users a project, product, or interesting thing. 

Users then interact by commenting and voting on the post. The Number of points on a post are cacluated by the number of upvotes minus the number of downvotes. 

## Analysis
Compare the different types of posts to determine:
 1. Do Ask HN or Show HN posts receive more comments on average?
 2. Do posts created at a certain time receive more comments on average?
 3. Do Ask HN or Show HN posts receive more points on average?
 4. Do posts created at a certain time receive more points on average?

In [1]:
## Importing the CSV file and converting it to list form
from csv import reader
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)

print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [2]:
#Remove headers
headers = hn[0]
hn = hn[1:]
print(headers)
print("\n")
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


## Ask HN or Show HN - Which post type receives more comments on average?
Because we want to understand if Ask HN or Show HN posts receive, on average, more comments, we will filter out and create a list with titles beginning with Ask Hn or Show HN only. 

I will do that using the string method startswith and looping through the HN data set.

In [3]:
#Seperate out Ask HN, Show Hn and other posts
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    title = title.lower()
    
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


In [4]:
print(ask_posts[:5])
print("\n")
print(show_posts[:5])

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 

## Average Number of Comments on Ask Posts

In [5]:
total_ask_comments = 0

for row in ask_posts:
    num_comm = int(row[4])
    total_ask_comments = total_ask_comments + num_comm

avg_ask_comments = (total_ask_comments / len(ask_posts))

print(avg_ask_comments)
    


14.038417431192661


## Average Number of Comments on Show Posts

In [6]:
total_show_comments = 0

for row in show_posts:
    num_comm_s = int(row[4])
    total_show_comments = total_show_comments + num_comm_s

avg_show_comments = (total_show_comments / len(show_posts))

print(avg_show_comments)

10.31669535283993


## Results of Average Number of Comments
Ask HN posts receive an average of 14 comments while Show HN posts receive an average of 10 comments. It will be beneficial to determine the times that generate the highest number of comments, on average for both types, if I would like to boost the engagement on either post type. 

# Finding the Amount of Ask Posts and Comments by Hour Created

Are posts created a certain time more likely to attact comments? To find this I will:

1. Calculate the amount of ask posts created in each hour of the day, along with the number of comments received. 
2. Calculate the average number of comments ask posts receive by hour created. 

In [7]:
#Calculate the amount of ask posts created during each hour of the day and the number of
#comments received.
import datetime as dt

result_list = []

for post in ask_posts:
    result_list.append([post[6], int(post[4])])

counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    created = row[0]
    comment = int(row[1])
    created_dt = dt.datetime.strptime(created, "%m/%d/%Y %H:%M")
    created_t = created_dt.strftime('%H')
    if created_t not in counts_by_hour:
        counts_by_hour[created_t] = 1
        comments_by_hour[created_t] = comment
    else: 
        counts_by_hour[created_t] += 1
        comments_by_hour[created_t] += comment

print(counts_by_hour)
print(comments_by_hour)
    
    

{'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}
{'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


In [8]:
# Calculate the average amount of comments 'Ask HN' posts created at each hour of the day 
# receive
avg_by_hour = []

for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])

print(avg_by_hour)

[['09', 5.5777777777777775], ['13', 14.741176470588234], ['10', 13.440677966101696], ['14', 13.233644859813085], ['16', 16.796296296296298], ['23', 7.985294117647059], ['12', 9.41095890410959], ['17', 11.46], ['15', 38.5948275862069], ['21', 16.009174311926607], ['20', 21.525], ['02', 23.810344827586206], ['18', 13.20183486238532], ['03', 7.796296296296297], ['05', 10.08695652173913], ['19', 10.8], ['01', 11.383333333333333], ['22', 6.746478873239437], ['08', 10.25], ['04', 7.170212765957447], ['00', 8.127272727272727], ['06', 9.022727272727273], ['07', 7.852941176470588], ['11', 11.051724137931034]]


In [9]:
# Sorting by number of comments
swap_avg_by_hour = []

for row in avg_by_hour:
    avg = row[1]
    hour = row[0]
    swap_avg_by_hour.append([avg,hour])

print(swap_avg_by_hour)

[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


In [10]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
print(sorted_swap[:5])


[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21']]


In [11]:
print("Top 5 Hours for Ask Posts Comments")

top_5 = []

for hour in sorted_swap[:5]:
    comms = hour[0]
    time = hour[1]
    time_dt = dt.datetime.strptime(time, "%H")
    time_t = time_dt.strftime("%H:%M")
    avg_com = "{0}: {1:.2f} average comments per post".format(time_t, comms)
    top_5.append(avg_com)

print(top_5)

Top 5 Hours for Ask Posts Comments
['15:00: 38.59 average comments per post', '02:00: 23.81 average comments per post', '20:00: 21.52 average comments per post', '16:00: 16.80 average comments per post', '21:00: 16.01 average comments per post']


## Results
Per the [documentation of the data set](https://www.kaggle.com/hacker-news/hacker-news-posts) the time was recorded on Eastern Standard Times. 

The hour that receives the highest number of average comments is 15:00, or 3:00pm est. That converts to 12 noon for Pacific Time. Posts at that time receive an average of 38.59 comments. That is followed by 2:00am est (11pm pt) with 23.81 average comments, which is 14.78 less comments compared to 15:00. 

# Finding the Amount of Show Posts and Comments by Hour Created

Are share posts created a certain time more likely to attact comments? While Show Posts do tend to attact less comments, I may still want to know the best time to post my Show HN conent.
To find this I will:

1. Calculate the amount of show posts created in each hour of the day, along with the number of comments received. 
2. Calculate the average number of comments show posts receive by hour created. 

In [12]:
#Calculate the amount of show posts created during each hour of the day and the number of
#comments received
import datetime as dt

result_list_show = []

for post in show_posts:
    result_list_show.append([post[6], int(post[4])])

counts_by_hour_s = {}
comments_by_hour_s = {}

for row in result_list_show:
    created = row[0]
    comment = int(row[1])
    created_dt = dt.datetime.strptime(created, "%m/%d/%Y %H:%M")
    created_t = created_dt.strftime('%H')
    if created_t not in counts_by_hour:
        counts_by_hour[created_t] = 1
        comments_by_hour[created_t] = comment
    else: 
        counts_by_hour[created_t] += 1
        comments_by_hour[created_t] += comment

print(counts_by_hour)
print(comments_by_hour)

{'09': 75, '13': 184, '10': 95, '14': 193, '16': 201, '23': 104, '12': 134, '17': 193, '15': 194, '21': 156, '20': 140, '02': 88, '18': 170, '03': 81, '05': 65, '19': 165, '01': 88, '22': 117, '08': 82, '04': 73, '00': 86, '06': 60, '07': 60, '11': 102}
{'09': 542, '13': 2199, '10': 1090, '14': 2572, '16': 2898, '23': 990, '12': 1407, '17': 2057, '15': 5109, '21': 2017, '20': 2334, '02': 1508, '18': 2401, '03': 708, '05': 522, '19': 1727, '01': 929, '22': 1049, '08': 657, '04': 584, '00': 934, '06': 539, '07': 566, '11': 1132}


In [13]:
# Calculate the average amount of comments 'Ask HN' posts created at each hour of the day 
# receive
avg_by_hour_s = []

for hr in comments_by_hour:
    avg_by_hour_s.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])

print(avg_by_hour_s)

[['09', 7.226666666666667], ['13', 11.951086956521738], ['10', 11.473684210526315], ['14', 13.32642487046632], ['16', 14.417910447761194], ['23', 9.51923076923077], ['12', 10.5], ['17', 10.658031088082902], ['15', 26.335051546391753], ['21', 12.929487179487179], ['20', 16.67142857142857], ['02', 17.136363636363637], ['18', 14.123529411764705], ['03', 8.74074074074074], ['05', 8.03076923076923], ['19', 10.466666666666667], ['01', 10.556818181818182], ['22', 8.965811965811966], ['08', 8.012195121951219], ['04', 8.0], ['00', 10.86046511627907], ['06', 8.983333333333333], ['07', 9.433333333333334], ['11', 11.098039215686274]]


In [14]:
# Sorting by number of comments
swap_avg_by_hour_s = []

for row in avg_by_hour_s:
    avg = row[1]
    hour = row[0]
    swap_avg_by_hour_s.append([avg,hour])


sorted_swap_s = sorted(swap_avg_by_hour_s, reverse=True)

print("Top 5 Hours for Show Posts Comments")

top_5 = []

for hour in sorted_swap_s[:5]:
    comms = hour[0]
    time = hour[1]
    time_dt = dt.datetime.strptime(time, "%H")
    time_t = time_dt.strftime("%H:%M")
    avg_com = "{0}: {1:.2f} average comments per post".format(time_t, comms)
    top_5.append(avg_com)

print(top_5)


Top 5 Hours for Show Posts Comments
['15:00: 26.34 average comments per post', '02:00: 17.14 average comments per post', '20:00: 16.67 average comments per post', '16:00: 14.42 average comments per post', '18:00: 14.12 average comments per post']


## Results

While there are less comments on average, 15:00, 3:00 est or 12:00 pt remains the hour with the highest number of comments for both ask and show posts. Similar to ask posts, 2:00 or 11:00 pt has the second highest average number of comments. However, Show Posts still have less comments at 26.34 and 17.14 respectively. 

Because certain hours seem to indicate higher level of engagement, I will next use the same methodology to calculate the average number of points per hour for both ask and show posts. 

# Average Number of Points on Ask HN and Show HN Posts

I reviewed the Hacker News [FAQ](https://news.ycombinator.com/newsfaq.html), which states that stories are ranked on the site using a basic algorithm that divides points by a power of the time since a story was submitted. Comments in threads are ranked the same way. 

Hacker News FAQ also states that comments to a post are closed after 2 weeks or if the post has been killed by a moderator or user flags. So I am curious about the number of points Ask HN and Show HN rececive. I would expect the number of points to be higher then comments and comments do drive rankings. 

Additionally, my hypothesis is that Ask HN inherently invites more comments through the nature of asking. Someone asks a question, others are inclined to answer. However, is it possible the Show HN gets more points, as users show their appreciation through an upvote but do not feel the need to comment? 

*Note: DataQuest cleaned this dataset to remove posts with 0 comments, so this is an analysis of posts with 0 comments only*

I want to note that average is not a perfect calculation for points or comments, however as I learn Python, I am applying what I know so far. Also, the content of the post can also impact the points scores and comments, but we do not have access to that data. That being said, a number of likes or upvotes as well as downvotes can provide a poster feedback about the content being posted, just like the average number of comments, and the hour a post obtains higher interaction can help a poster know when and what type of content could be successful.

So, although not a part of DataQuest's instructions, I am adding to my analysis as we have the data to do so. 

Next, I will perform a similar calculation to find the average number of points on Ask HN and Share HN posts. 

## Average Number of Points on an Ask Post

In [22]:
total_ask_points = 0

for row in ask_posts:
    points = int(row[3])
    total_ask_points = total_ask_points + points

avg_ask_points = (total_ask_points / len(ask_posts))

print(avg_ask_points)

15.061926605504587


## Average Number of Points on a Show Post

In [15]:
total_show_points = 0

for row in show_posts:
    points = int(row[3])
    total_show_points = total_show_points + points

avg_show_points = (total_show_points / len(show_posts))

print(avg_show_points)

27.555077452667813


## Results for Average Number of Points
My hypothesis proved true, as Show HN posts have a higher number of average points (27) than Ask HN posts (15). As noted above, this dataset has been cleaned so that all posts with 0 comments have been removed, so the dataset may look different, however since the goal is to increase engagement, having both comments and points shows higher engagement. By asking, a user will comment to provide the answer but may or may not vote. By showing, it might be duplicative for a user to upvote and comment on the post, unless providing feedback.

## Finding the Amount of Ask Posts and Points by Hour Created

In [16]:
import datetime as dt

pts_result_list = []

for row in ask_posts:
    pts_result_list.append([row[6], int(row[3])])

pts_counts_by_hour = {}
points_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for row in pts_result_list:
    date = row[0]
    point = row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time in pts_counts_by_hour:
        points_by_hour[time] += point
        pts_counts_by_hour[time] += 1
    else:
        points_by_hour[time] = point
        pts_counts_by_hour[time] = 1
        
print(points_by_hour)
        

{'09': 329, '13': 2062, '10': 1102, '14': 1282, '16': 2522, '23': 581, '12': 782, '17': 1941, '15': 3479, '21': 1721, '20': 1151, '02': 793, '18': 1741, '03': 374, '05': 552, '19': 1513, '01': 700, '22': 511, '08': 515, '04': 389, '00': 451, '06': 591, '07': 361, '11': 825}


In [17]:
avg_pts_by_hour = []

for hr in points_by_hour:
    avg_pts_by_hour.append([hr, points_by_hour[hr] / pts_counts_by_hour[hr]])

print(avg_pts_by_hour)

[['09', 7.311111111111111], ['13', 24.258823529411764], ['10', 18.677966101694917], ['14', 11.981308411214954], ['16', 23.35185185185185], ['23', 8.544117647058824], ['12', 10.712328767123287], ['17', 19.41], ['15', 29.99137931034483], ['21', 15.788990825688073], ['20', 14.3875], ['02', 13.672413793103448], ['18', 15.972477064220184], ['03', 6.925925925925926], ['05', 12.0], ['19', 13.754545454545454], ['01', 11.666666666666666], ['22', 7.197183098591549], ['08', 10.729166666666666], ['04', 8.27659574468085], ['00', 8.2], ['06', 13.431818181818182], ['07', 10.617647058823529], ['11', 14.224137931034482]]


In [18]:
swap_avg_by_hour_pts = []

for row in avg_pts_by_hour:
    swap_avg_by_hour_pts.append([row[1], row[0]])

sorted_swap_pts = sorted(swap_avg_by_hour_pts, reverse=True)

print("Top 5 Hours for 'Ask HN' Points")
for avg, hr in sorted_swap_pts[:5]:
    print(
        "{}: {:.2f} average points per ask post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"), avg
        )
    )

Top 5 Hours for 'Ask HN' Points
15:00: 29.99 average points per ask post
13:00: 24.26 average points per ask post
16:00: 23.35 average points per ask post
17:00: 19.41 average points per ask post
10:00: 18.68 average points per ask post


The top hour for points, 15:00 remains the same for comments and posts so far. 


# Finding the Amount of Show Posts and Points by Hour Created

In [19]:
import datetime as dt

pts_result_list_s = []

for row in show_posts:
    pts_result_list_s.append([row[6], int(row[3])])

pts_counts_by_hour_s = {}
points_by_hour_s = {}
date_format = "%m/%d/%Y %H:%M"

for row in pts_result_list_s:
    date = row[0]
    point = row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time in pts_counts_by_hour_s:
        points_by_hour_s[time] += point
        pts_counts_by_hour_s[time] += 1
    else:
        points_by_hour_s[time] = point
        pts_counts_by_hour_s[time] = 1
        
print(points_by_hour_s)

{'14': 2187, '22': 1856, '18': 2215, '07': 494, '20': 1819, '05': 104, '16': 2634, '19': 1702, '15': 2228, '03': 679, '17': 2521, '06': 375, '02': 340, '13': 2438, '08': 519, '21': 866, '04': 386, '11': 1480, '12': 2543, '23': 1526, '09': 553, '01': 700, '10': 681, '00': 1173}


In [20]:
avg_pts_by_hour_s = []

for hr in points_by_hour_s:
    avg_pts_by_hour_s.append([hr, points_by_hour_s[hr] / pts_counts_by_hour_s[hr]])

print(avg_pts_by_hour_s)

[['14', 25.430232558139537], ['22', 40.34782608695652], ['18', 36.31147540983606], ['07', 19.0], ['20', 30.316666666666666], ['05', 5.473684210526316], ['16', 28.322580645161292], ['19', 30.945454545454545], ['15', 28.564102564102566], ['03', 25.14814814814815], ['17', 27.107526881720432], ['06', 23.4375], ['02', 11.333333333333334], ['13', 24.626262626262626], ['08', 15.264705882352942], ['21', 18.425531914893618], ['04', 14.846153846153847], ['11', 33.63636363636363], ['12', 41.68852459016394], ['23', 42.388888888888886], ['09', 18.433333333333334], ['01', 25.0], ['10', 18.916666666666668], ['00', 37.83870967741935]]


In [21]:
swap_avg_by_hour_pts_s = []

for row in avg_pts_by_hour_s:
    swap_avg_by_hour_pts_s.append([row[1], row[0]])

sorted_swap_pts_s = sorted(swap_avg_by_hour_pts_s, reverse=True)

print("Top 5 Hours for 'Show HN' Points")
for avg, hr in sorted_swap_pts_s[:5]:
    print(
        "{}: {:.2f} average points per show post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"), avg
        )
    )

Top 5 Hours for 'Show HN' Points
23:00: 42.39 average points per show post
12:00: 41.69 average points per show post
22:00: 40.35 average points per show post
00:00: 37.84 average points per show post
18:00: 36.31 average points per show post


As discovered earlier, Show HN posts receive more points then Ask HN posts. It also appears that 23:00, or 11PM est and 8PM pt receive the most points (42.39) followed by noon est or 9am pt (41.69).

# Conclusion

Below is a table summarizing the hours with the highest number of comments and points for both Ask HN and Post HN posts. 

| Post Type | Time & Comments | Time & Points |
|-----------|-----------------|---------------|
| Ask HN    | 15:00 38.59     | 15:00 29.99   |
| Post HN   | 15:00 26.34     | 23:00 42.39   |
|           |                 |               |

These are the hours with the second highest number of comments and points:

| Post Type | Time & Comments | Time & Points |
|-----------|-----------------|---------------|
| Ask HN    | 02:00 23.81     | 13:00 24.26   |
| Post HN   | 02:00 17.14     | 12:00 41.69   |
|           |                 |               |

Depending on your goal, higher points or higher comments, this will help you determine when to post. Engagement appears to be high at 15:00 or 3:00pm EST and 12:00pm PT where you could get both the highest number of commments for both Ask HN and Post HN posts, as well as the highest number of points for Ask HN. 

However, if you are looking for higher points, a Show HN post will generate more points. 4 of 5 top hours for Show HN have average points per hour greater than 38.59, the top number of Ask HN comments. 

One possible strategy to get the most exposure is to first post an Ask HN post asking for feedback on a project and once finalizing the project post a Show HN post. Otherwise, the results are helpful when determining if you would like a high number of comments or higher points. 

When determining what time to post, please be sure to convert to your local time zone. 

A few more recommendations to consider when posting:
 - What content generates more comments and points
 
Happy Posting!